client/src/legacy/GridField.js

```diff
@@ -180,6 +199,45 @@ $.entwine('ss', function($) {
       if (hasLegacyFilterHeader) {
         this.find('.sortable-header th:last').html(content);
       }
+    },
+
+    keepStateInHistory: function() {
+      const newSchema = $(this).find('.gridfield-actionmenu__container').data('schema');
+      if (newSchema && newSchema.length > 0) {
+        newSchema.filter( e => {
+          if (e.type === 'link') {
+            const reqString = this.buildURLString(e.url)
+            const searchParam =  reqString ? '?' + reqString.join('&') : '';
+            const historyState = $.extend({}, { page: window.location.pathname + searchParam }, this.getState());
+            history.replaceState(historyState, '', window.location.pathname + searchParam);
```

Comment:
```suggestion
            window.ss.router.replace(window.location.pathname + searchParam, undefined, undefined, false);
```
Turns out we have to use the special custom page router that Silverstripe admin uses... see https://github.com/visionmedia/page.js/blob/master/page.js#L678 for what this is actually doing.